### DISCUSSION TITLE
The type initializer for 'Ganss.Xss.HtmlSanitizer' threw an exception.

### DISCUSSION THREAD

**Discussion by byap01** [mganss/HtmlSanitizer#407] [Q&A]
The type initializer for 'Ganss.Xss.HtmlSanitizer' threw an exception.

Inner Exception 1:
FileLoadException: Could not load file or assembly 'AngleSharp, Version=0.16.1.0, Culture=neutral, PublicKeyToken=e83494dcdc6d31ea' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)

HtmlSanitizer packagereference?
![image](https://user-images.githubusercontent.com/101874426/200586381-5bad363a-f196-4920-a71a-024d6da9f79e.png)

AngleSharp.Css (0.16.4) package reference
![image](https://user-images.githubusercontent.com/101874426/200586440-4ccd63ba-dcee-4e99-a231-b4ceced0a9f3.png)

Please help on how to do upgrade this?

**Comment by mganss** [comment]
Where's that second package reference line from?
AngleSharp.Css 0.16.4 has a dependency on [AngleSharp (>= 0.16.1)](https://www.nuget.org/packages/AngleSharp.Css/0.16.4#dependencies-body-tab)

**Reply by byap01** [reply]
as my error says it keeps on looking for 0.16.1 which I don't have, since angleSharp is upgraded to 0.17.
I get that documentation says >=0.16.1 but keeps on looking for 0.16.1

a snippet from project.assets.json says:
      "HtmlSanitizer/8.0.601": {
        "type": "package",
        "dependencies": {
          "AngleSharp": "[0.17.1]",
          "AngleSharp.Css": "[0.16.4]",
          "System.Collections.Immutable": "6.0.0"
        },
        
		"AngleSharp/0.17.1": {
        "type": "package",
        "dependencies": {
          "System.Buffers": "4.5.1",
          "System.Text.Encoding.CodePages": "6.0.0"
        },
        "compile": {
          "lib/net472/AngleSharp.dll": {
            "related": ".xml"
          }
        },
        "runtime": {
          "lib/net472/AngleSharp.dll": {
            "related": ".xml"
          }
        }
      },
      
	  "AngleSharp.Css/0.16.4": {
        "type": "package",
        "dependencies": {
          "AngleSharp": "0.16.1"
        },
        "compile": {
          "lib/net472/AngleSharp.Css.dll": {
            "related": ".xml"
          }
        },

**Reply by mganss** [reply]
Where is the `<PackageReference Include="AngleSharp" Version="[0.16.1]" />` from?

**Reply by byap01** [reply]
From AngleSharp.css(0.16.4) which is used by htmlsanitizer

**Reply by mganss** [reply]
Can't see this. AFAICT, AngleSharp.Css depends on AngleSharp >= 0.16.1:
```xml
<PackageReference Include="AngleSharp" Version="0.16.1" />
```

https://github.com/AngleSharp/AngleSharp.Css/blob/a754c9adb1d678341e9f489674a8ad7a7199eae8/src/AngleSharp.Css/AngleSharp.Css.csproj#L24

I have a feeling this has something to do with the NuGet package cache. Perhaps try clearing it: https://learn.microsoft.com/en-us/nuget/consume-packages/managing-the-global-packages-and-cache-folders#clearing-local-folders

**Reply by byap01** [reply]
I did the cache clearing already and still have this issue. There were already anglesharp.css version (0.17) but still on beta mode.
any chance we can update the package to use the latest beta version?

**Reply by mganss** [reply]
I would try to find the root cause of this issue in your setup. Does it work if you create a simple console application from scratch?

**Reply by tiesont** [reply]
> I did the cache clearing already and still have this issue. There were already anglesharp.css version (0.17) but still on beta mode. any chance we can update the package to use the latest beta version?

There's a reason why HtmlSanitizer is usually behind AngleSharp: https://github.com/mganss/HtmlSanitizer/wiki/Frequently-Asked-Questions#why-is-the-anglesharp-dependency-pinned-to-a-version-less-than-the-head-revision

So unless @mganss pushes a pre-release version, I wouldn't expect to see an update until 0.17 is no longer beta.

**Reply by byap01** [reply]
> > I did the cache clearing already and still have this issue. There were already anglesharp.css version (0.17) but still on beta mode. any chance we can update the package to use the latest beta version?

Yes, I did a simple application from scratch and it did work.

**Reply by mganss** [reply]
Have you tried deleting the .vs, bin, and obj folders in your original project?

**Reply by byap01** [reply]
What I did for now just assembly binding which has work so far.

**Reply by msgill1989** [reply]
I'm also getting this same error. When I use HtmlSanitizer version 8.0.645 it gives error for AngleSharp v 0.17.0 and when I use Version 8.0.601 its giving error for AngleSharp 0.16.1.

My package reference is 
    <PackageReference Include="HtmlSanitizer">
      <Version>8.0.601</Version>
    </PackageReference>

I did not install AngleSharp nuget package, but I checked the dependency is there in bin folder.

I tried reinstalling the package, removing nuget cache and deleting .vs, bin and Obj folder.

**Reply by msgill1989** [reply]
@mganss Were you able to get the root-cause of the issue?

**Reply by mganss** [reply]
@msgill1989 Unfortunately, no. I can't reproduce the issue.

**Reply by mbowles-bhill** [reply]
Is there still no resolution on this? I am running into this issue.
![image](https://github.com/user-attachments/assets/254f5a1e-aab8-4974-8ea8-ca9908a274ee)
We have AngleSharp 0.17.1 installed on that project so I don't know where its getting 0.17.0

**Reply by mbowles-bhill** [reply]
As a workaround for any future readers, I used bindingredirect in my web.config to force the version of AngleSharp that it wanted and it works. There's still an issue somewhere but this got past the issue for now
